src/app/grid/grid-save-state/state.directive.ts

```diff
@@ -85,72 +95,56 @@ export class IgxGridStateDirective implements AfterViewInit {
-              }
-              // we pass an array of IFilteringExpression to the createExpressionsTree
-              const columnFilteringExpressionsTree = this.createExpressionsTree(columnsFiltOperands, filtOperand);
-              gridFilteringExpressionsTree.filteringOperands.push(columnFilteringExpressionsTree);
-            }
+            const filterTree = this.createExpressionsTreeFromObject(this.gridState.filtering);
+            this.grid.filteringExpressionsTree = filterTree;
+        }
 
-            this.grid.filteringExpressionsTree = gridFilteringExpressionsTree;
-          }
+        // restore advanced filtering
+        if (this.advancedFiltering && this.gridState.advancedFiltering) {
+            const advFilterTree = this.createExpressionsTreeFromObject(this.gridState.advancedFiltering);
+            this.grid.advancedFilteringExpressionsTree = advFilterTree;
+        }
 
         // restore paging
         if (this.paging && this.gridState.paging) {
-          if (this.grid.perPage !== this.gridState.paging.recordsPerPage) {
-            this.grid.perPage = this.gridState.paging.recordsPerPage;
-            this.grid.cdr.detectChanges();
-          }
-          if (this.grid.page !== this.gridState.paging.index) {
-            this.grid.paginate(this.gridState.paging.index);
-          }
+            if (this.grid.perPage !== this.gridState.paging.recordsPerPage) {
+                this.grid.perPage = this.gridState.paging.recordsPerPage;
+                this.grid.cdr.detectChanges();
+            }
+            if (this.grid.page !== this.gridState.paging.index) {
+                this.grid.paginate(this.gridState.paging.index);
+            }
         }
 
         // restore sorting
         if (this.sorting && this.gridState.sorting) {
-          const strategy = DefaultSortingStrategy.instance();
-          this.gridState.sorting.forEach((expr) => expr.strategy = strategy);
-          this.grid.sortingExpressions = this.gridState.sorting;
+            const strategy = DefaultSortingStrategy.instance();
+            this.gridState.sorting.forEach((expr) => expr.strategy = strategy);
+            this.grid.sortingExpressions = this.gridState.sorting;
         }
 
         // restore selection
         if (this.selection && this.gridState.selection) {
-          this.grid.selectRows(this.gridState.selection);
+            this.grid.selectRows(this.gridState.selection);
         }
     }
 
-    public storeState(action: string, args: any) {
+    public storeState(action: string, args: any, stringifyCb?: (key, val) => any) {
         if (this[action]) {
             const actionKey = action + "-" + this.grid.id;
-            window.localStorage.setItem(actionKey, JSON.stringify(args));
+            if (stringifyCb) {
+                window.localStorage.setItem(actionKey, JSON.stringify(args, stringifyCb));
+            } else {
+                window.localStorage.setItem(actionKey, JSON.stringify(args));
+            }
```

Comment:
There is no need to check `stingifyCb`, you can always call `window.localStorage.setItem(actionKey, JSON.stringify(args, stringifyCb));`. Same is valid for `parseCb` below.

Reply:
Done